File: demo/app/scripts/build-embedded-python.ts
// scripts/build-embedded-python.ts
// Clean script to build the Python wheel and embed it into the ElectroBun resources.
//
// Run with:
//   bun scripts/build-embedded-python.ts
//
// This will:
// 1. Build the wheel from the monorepo root using uv
// 2. Bundle a standalone Python + the wheel into demo/app/resources/python
// 3. The resulting python binary can be used by python-bridge.ts
//
// The ElectroBun app (when not in DEV_PYTHON_BIN mode) will spawn this embedded Python.

import { $ } from "bun";
import { existsSync, mkdirSync } from "fs";
import { join, resolve } from "path";
import { parseArgs } from "util";

const { values } = parseArgs({
  args: Bun.argv.slice(2),
  options: {
    platform: { type: "string" },
    clean: { type: "boolean" },
  },
});

const REPO_ROOT = resolve(import.meta.dir, "../../..");
const DEMO_APP_DIR = resolve(import.meta.dir, "..");
const DIST_DIR = join(REPO_ROOT, "dist");
const RESOURCES_PYTHON = join(DEMO_APP_DIR, "resources/python");

console.log("==> Building embedded Python for ElectroBun demo");
console.log(`    Repo root: ${REPO_ROOT}`);
console.log(`    Target resources: ${RESOURCES_PYTHON}`);

if (values.clean) {
  console.log("==> Cleaning previous resources/python");
  await $`rm -rf ${RESOURCES_PYTHON}`.quiet().nothrow();
}

// 1. Build the wheel
console.log("\n==> Building wheel with uv...");
await $`cd ${REPO_ROOT} && uv build --wheel --out-dir ${DIST_DIR}`;

// 2. Find the wheel
const wheelFiles = (await $`ls ${DIST_DIR}/corpora_py-*.whl 2>/dev/null || ls ${DIST_DIR}/*.whl`.text()).trim().split("\n").filter(Boolean);
if (wheelFiles.length === 0) {
  console.error("No wheel found in dist/. Did uv build succeed?");
  process.exit(1);
}

const wheelPath = wheelFiles[0];
console.log(`    Found wheel: ${wheelPath}`);

// 3. Run the bundler (it will create resources/python)
console.log("\n==> Bundling standalone Python + wheel...");
const bundleScript = join(DEMO_APP_DIR, "scripts/bundle-python.ts");
const bundleArgs = [wheelPath];
if (values.platform) {
  bundleArgs.push("--platform", values.platform);
}

await $`cd ${DEMO_APP_DIR} && bun ${bundleScript} ${bundleArgs}`;

console.log("\n✅ Done!");
console.log(`   Embedded Python is at: ${RESOURCES_PYTHON}/bin/python3`);
console.log("\n   To use it locally (without Docker dev override):");
console.log("     unset DEV_PYTHON_BIN DEV_PYTHON_HOME");
console.log("     bun run dev     # or whatever launches the ElectroBun app");
console.log("\n   The python-bridge.ts will automatically use this embedded Python.");

File: scripts/build_embedded_python.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.13"
# ///
"""
Build the wheel for this project and embed it into the ElectroBun demo app's
standalone Python runtime.

Python equivalent of demo/app/scripts/build-embedded-python.ts

Usage:
    uv run python scripts/build_embedded_python.py [--platform ...] [--clean]
"""

import argparse
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DIST_DIR = ROOT / "dist"
DEMO_RESOURCES = ROOT / "demo/app/resources/python"
BUNDLE_SCRIPT = ROOT / "scripts/bundle_python.py"


def run(cmd: list[str], cwd: Path = ROOT):
    print(f"$ {' '.join(str(c) for c in cmd)}")
    subprocess.run(cmd, cwd=cwd, check=True)


def find_wheel() -> Path:
    wheels = sorted(
        DIST_DIR.glob("corpora_py-*.whl"), key=lambda p: p.stat().st_mtime, reverse=True
    )
    if not wheels:
        wheels = sorted(
            DIST_DIR.glob("*.whl"), key=lambda p: p.stat().st_mtime, reverse=True
        )
    if not wheels:
        raise SystemExit("No wheel found in dist/. Did `uv build` succeed?")
    return wheels[0]


def parse_args():
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument(
        "--platform",
        help="Target platform for the standalone Python (passed to bundle script)",
    )
    p.add_argument(
        "--clean", action="store_true", help="Remove previous resources/python first"
    )
    p.add_argument(
        "--full",
        action="store_true",
        help="Install with [full] extra (includes text-fabric for conversions)",
    )
    return p.parse_args()


def main():
    args = parse_args()

    python_bin = DEMO_RESOURCES / "bin" / "python3"

    if python_bin.exists() and not args.clean:
        print("==> Embedded Python runtime for demo already present (skipping).")
        print(f"   {python_bin}")
        return

    print("==> Building embedded Python for ElectroBun demo")
    print(f"    Repo root: {ROOT}")
    print(f"    Target:    {DEMO_RESOURCES}")

    if args.clean:
        print("==> Cleaning previous resources/python")
        if DEMO_RESOURCES.exists():
            import shutil

            shutil.rmtree(DEMO_RESOURCES)

    # 1. Build wheel
    print("\n==> Building wheel with uv...")
    DIST_DIR.mkdir(parents=True, exist_ok=True)
    run(["uv", "build", "--wheel", f"--out-dir={DIST_DIR}"])

    wheel = find_wheel()
    print(f"    Found wheel: {wheel.name}")

    wheel_spec = f"{wheel}[full]" if args.full else str(wheel)
    print(f"\n==> Bundling standalone Python + wheel{'[full]' if args.full else ''}...")

    bundle_cmd = [sys.executable, str(BUNDLE_SCRIPT), wheel_spec]
    if args.platform:
        bundle_cmd += ["--platform", args.platform]
    bundle_cmd += ["--dest-dir", str(DEMO_RESOURCES)]

    run(bundle_cmd)

    print("\n✅ Done!")
    print(f"   Embedded Python is at: {DEMO_RESOURCES}/bin/python3")
    print("\n   To use it locally (without Docker dev override):")
    print("     unset DEV_PYTHON_BIN DEV_PYTHON_HOME")
    print("     bun run dev     # or the equivalent ElectroBun launch command")
    print("\n   The python-bridge will automatically use this embedded Python.")


if __name__ == "__main__":
    main()
